File: Ex0207_ReverseLinkedList/ReverseLinkedList.c
#include <stdio.h>
#include <stdlib.h>

typedef int datatype;

struct node
{
	datatype data;
	struct node* next;
};

typedef struct node* PtrToNode;
typedef PtrToNode List, Position;

int IsEmpty(List L)
{
	return L->next == NULL;
}

int IsLast(Position P)
{
	return P->next == NULL;
}

// Generate a new empty list
List MakeEmpty(void)
{
	List L = (List)malloc(sizeof(struct node));
	if (L == NULL)
	{
		printf("Error in allocate space.\n");
		exit(1);
	}
	L->next = NULL;

	return L;
}

// Delete list elements and make it an empty list
void DeleteList(List L)
{
	Position P, tmp;
	P = L->next;
	L->next = NULL;
	while (P != NULL)
	{
		tmp = P->next;
		free(P);
		P = tmp;
	}
}

void ShowList(List L)
{
	Position P = L->next;
	while (P != NULL)
	{
		printf("%d, ", P->data);
		P = P->next;
	}
	printf("\n");
}

// Insert a node after position p
void Insert(datatype x, List L, Position P)
{
	Position TmpCell;

	TmpCell = malloc(sizeof(struct node));
	if (TmpCell == NULL)
	{
		printf("Error in allocate space.\n");
		exit(1);
	}
	
	TmpCell->data = x;
	TmpCell->next = P->next;
	P->next = TmpCell;
}

void BuildList(List L, const datatype input_array[], int size)
{
	Position last = L;
	for (int i = 0; i < size; i++)
	{
		Insert(input_array[i], L, last);
		last = last->next;
	}
}

void Reverse(List L)
{
	if (IsEmpty(L) || IsLast(L->next))
	{
		printf("Do not need to reverse the list.\n");
		return;
	}

	Position current, nextnode;
	current = nextnode = L->next->next;
	L->next->next = NULL;
	while (current != NULL)
	{
		nextnode = current->next;
		current->next = L->next;
		L->next = current;
		current = nextnode;
	}
}

int main(void)
{
	List L = MakeEmpty();
	datatype input_array[] = { 1,9,3,5,4,2,7,4,5 };
	BuildList(L, input_array, sizeof(input_array) / sizeof(input_array[0]));
	printf("List before reversion:\n");
	ShowList(L);

	Reverse(L);

	printf("List after reversion:\n");
	ShowList(L);
	DeleteList(L);
	free(L);
	L = NULL;
}

File: Ex0207_ReverseLinkedList/ReverseLinkedList.java

public class ReverseLinkedList {
    public static void main(String[] args) {
        MyLinkedList<Integer> list = new MyLinkedList<>(
                new Integer[]{1,9,3,5,4,2,7,4,5});

        System.out.println("List before reversion:");
        list.show();

        list.reverse();

        System.out.println("List after reversion:");
        list.show();
    }
}

class MyLinkedList<E> {
    node<E> head = new node<>();

    public static class node<E> {
        private E data = null;
        private node<E> next = null;

        public node() {}

        public node(E data) {
            setData(data);
        }

        public node(E data, node<E> next) {
            this(data);
            setNext(next);
        }

        public E getData() {
            return data;
        }

        public void setData(E data) {
            this.data = data;
        }

        public node<E> getNext() {
            return next;
        }

        public boolean isLast() {
            return getNext() == null;
        }

        protected void setNext(node<E> next) {
            this.next = next;
        }
    }

    public MyLinkedList() { }

    public MyLinkedList(E[] inputArray) {
        buildList(inputArray);
    }

    public boolean isEmpty() {
        return head.isLast();
    }

    public void makeEmpty() {
        head.setNext(null);
    }

    public void show() {
        node<E> p = head.getNext();

        while(p != null) {
            System.out.print(p.getData() + ", ");
            p = p.getNext();
        }
        System.out.println();
    }

    public void insert(E x, node<E> p) {
        node<E> tmpNode = new node<>(x, p.next);
        p.setNext(tmpNode);
    }

    public void buildList(final E[] inputArray) {
        makeEmpty();
        node<E> last = head;
        for (E e: inputArray) {
            insert(e, last);
            last = last.next;
        }
    }

    public void reverse() {
        if (isEmpty() || head.getNext().isLast()) {
            System.out.println("Do not need to reverse the list.");
            return;
        }

        node<E> current, nextNode;
        current = head.getNext().getNext();
        head.getNext().setNext(null);

        while (current != null) {
            nextNode = current.getNext();
            current.setNext(head.getNext());
            head.setNext(current);
            current = nextNode;
        }
    }

}
